### ISSUE TITLE
Inserting special tokens

### ISSUE THREAD

**Issue by david-wb** [JohnGiorgi/DeCLUTR#195]
Say I want to insert special tokens into a piece of text to help the model distinguish certain features. 

With BERT, I can use tokens like `[unused0], [unused1], ...`. Are there similar tokens I can use with DeCLUTR? 

For context, I'm using special tokens to delineate the boundaries of named entity mentions.
For example, inserting special tokens into
```
Jim bought 300 shares of Acme Corp. in 2006.
```
yields
```
Jim bought 300 shares of [unused0] Acme Corp. [unused1] in 2006.
```

How can I do this with DeCLUTR? Thanks!

**Comment by JohnGiorgi** [comment]
This depends quite a bit on the details of your setup. I will assume you are doing everything from [HuggingFace's Transformers](https://huggingface.co/transformers/index.html). Then the first step is to take advantage of the `additional_special_tokens` kwarg of the [Transformers tokenizers](https://huggingface.co/transformers/main_classes/tokenizer.html). Something like:

```python
from transformers import AutoTokenizer

tokens_to_add = [
    "[unused0]",
    "[unused1]",
]

tokenizer = AutoTokenizer.from_pretrained("johngiorgi/declutr-small", additional_special_tokens=tokens_to_add)
```

Now, the tokenizer will not split these tokens up,

```python
text = "Jim bought 300 shares of [unused0] Acme Corp. [unused1] in 2006."
tokenizer.tokenize(text)
>>> ['Jim',
 'Ġbought',
 'Ġ300',
 'Ġshares',
 'Ġof',
 'Ġ',
 '[unused0]',
 'ĠAc',
 'me',
 'ĠCorp',
 '.',
 'Ġ',
 '[unused1]',
 'Ġin',
 'Ġ2006',
 '.']
```

The next step is to resize the models embedding. Here is some code to do that, borrowed from [AllenNLP](https://github.com/allenai/allennlp/blob/8fbc9728e99a85483b8a061b246c6547bab15e40/allennlp/modules/token_embedders/pretrained_transformer_embedder.py#L108-L118):

```python
from transformers import AutoModel

model = AutoModel.from_pretrained("johngiorgi/declutr-small")

try:
    if model.get_input_embeddings().num_embeddings != len(tokenizer):
        model.resize_token_embeddings(len(tokenizer))
except NotImplementedError:
    # Can't resize for transformers models that don't implement base_model.get_input_embeddings()
    logger.warning(
        "Could not resize the token embedding matrix of the transformer model. "
        "This model does not support resizing."
    )
```

Of course, you now need to train this thing to make use of your new special tokens.

If you want to do this in AllenNLP (which DeCLUTR is implemented in), the steps are the same, but you will likely do this via AllenNLPs config files.

**Comment by david-wb** [comment]
Thanks for explaining. It seems RoBERTa does not have additional unused tokens built in to the vocab, and I do not want to retrain from scratch, so I will try using some other delimiter like a simple `|`.

**Comment by JohnGiorgi** [comment]
Maybe I am misunderstanding, but I don't think you would need to train from scratch in either case.

I think regardless of the strategy (add additional special tokens or use the unused ones) you will have to fine-tune the model so it learns to use your special tokens effectively / in the way that you expect.

Hope that helps.

**Comment by david-wb** [comment]
Oh so resizing the model input embeddings would not force a re-train from scratch? If so I will likely try that. Thank you.

**Comment by JohnGiorgi** [comment]
I can't think of why it would, but I would double-check to be sure! Might be worth it to dig into the [`resize_token_embeddings`](https://huggingface.co/transformers/main_classes/model.html#transformers.PreTrainedModel.resize_token_embeddings) function.
